sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
-        } catch (...) {
-            _all->erase(sst);
-            throw;
+sstable_set_version_reference::sstable_set_version_reference(const sstable_set_version_reference& ref) 
+    : _p(ref._p ? new sstable_set_version(ref._p) : nullptr) {
+        if (_p) {
+            _p->_reference_count++;
         }
-    } catch (...) {
-        _impl->erase(sst);
-        throw;
+}
+
+sstable_set_version_reference::sstable_set_version_reference(sstable_set_version_reference&& ref) : _p(ref._p) {
+    ref._p = nullptr;
+}
+
+sstable_set_version_reference& sstable_set_version_reference::operator=(const sstable_set_version_reference& ref) {
+    if (this != &ref) {
+        *this = sstable_set_version_reference(ref);
     }
+    return *this;
 }
 
-void
-sstable_set::erase(shared_sstable sst) {
-    _impl->erase(sst);
-    _all->erase(sst);
-    _all_runs[sst->run_identifier()].erase(sst);
+sstable_set_version_reference& sstable_set_version_reference::operator=(sstable_set_version_reference&& ref) {
+    if (this != &ref) {
+        this->~sstable_set_version_reference();
+        new (this) sstable_set_version_reference(std::move(ref));
+    }
+    return *this;
 }
 
-sstable_set::~sstable_set() = default;
+sstable_list::sstable_list(std::unique_ptr<sstable_set_impl> impl, schema_ptr s) 
+    : _version(new sstable_set_version(std::move(impl), std::move(s))) {
+}
 
-sstable_set::incremental_selector::incremental_selector(std::unique_ptr<incremental_selector_impl> impl, const schema& s)
-    : _impl(std::move(impl))
-    , _cmp(s) {
+sstable_list::sstable_list(const sstable_list& sstl)
+    : _version(sstl._version) {
+}
+
+sstable_list::sstable_list(sstable_list&& sstl) noexcept = default;
+
+sstable_list& sstable_list::operator=(const sstable_list& sstl) {
+    if (this != &sstl) {
+        *this = sstable_list(sstl);
+    }
+    return *this;
+}
+
+sstable_list& sstable_list::operator=(sstable_list&& sstl) noexcept {
+    if (this != &sstl) {
+        this->~sstable_list();
+        new (this) sstable_list(std::move(sstl));
+    }
+    return *this;
+}
+
+sstable_list::const_iterator::const_iterator(std::set<shared_sstable>::const_iterator it, const sstable_set_version* ver)
+    : _it(std::move(it))
+    , _ver(ver)
+{
+    while (_it != _ver->all().end() && !_ver->contains(*_it)) {
+        _it++;
+    }
+}
+
+sstable_list::const_iterator& sstable_list::const_iterator::operator++() {
+    if (_it != _ver->all().end()) {
+        _it++;
+    }
+    while (_it != _ver->all().end() && !_ver->contains(*_it)) {
+        _it++;
+    }
+    return *this;
+}
+
+sstable_list::const_iterator sstable_list::const_iterator::operator++(int) {
+    auto it = *this;
+    operator++();
+    return it;
+}
+
+const shared_sstable& sstable_list::const_iterator::operator*() const {
+    return *_it;
+}
+
+bool sstable_list::const_iterator::operator==(const const_iterator& it) const {
+    return _it == it._it && _ver == it._ver;
```

Comment:
In theory it is illegal to compare iterators that come from different containers (and here different versions are logically equivalent to different containers).

Reply:
Should we assert that then? Or could we allow that and check first if versions match  (`_ver == it._ver && _it == it._it`) so that we don't compare `_it` if they come from different containers?